1268 indices class

hiker opened the pull request:
This fixed #1268 and #1269 (the latter was so easy to fix I didn't bother putting this as a separate PR).

hiker commented:
That is ready for a first review. Note that I decided not to expose the ComponentIndices when adding variable access information, I think with the support for simplified parameter in the constructor  I am quite happy with what PSyclone needs to do (compared with creating a separate class). Additionally, we can now handle component like `field%whole%x_start` (or so) without parameter, the `add_access` method will automatically create a ComponentIndices with `[ [], [], [] ]` for this scalar variable, and also `field%data` with ['i', 'j'] as parameter will correctly create `[ [], [i,j] ]`.

arporter commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -165,59 +243,15 @@ def array_access_parallelisable(self, loop_variable, var_info):
+        old_num_messages = len(self.get_all_messages())
+        all_indices = \
+            self.array_accesses_consistent(signature, loop_variable, var_info)
+        if len(self.get_all_messages()) > old_num_messages:
```
Not something for this PR but it feels like the dependency checking should be configurable - i.e. it would be good if we could set the 'log level' to abort with an exception on first error. We could then use try..except which is better than counting error messages.

hiker replied:
Yes, that code is particularly ugly :( My idea that the dependency tool leaves it to the caller to decide how to handle error (and especially allowing the dependency tools to collect as many errors as possible, so that the user can look at them all at once and fix them at once) breaks down the moment we have a function that is used internally and externally :( To you want me to open a ticket for that?

arporter replied:
Yes please. Maybe changing it to do logging might help? Also, as I say earlier, I think the `array_accesses_consistent` method should really wrap some of this and just return True or False.

hiker replied:
Fixed when refactoring `array_access_consistent`.

arporter commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -240,7 +274,7 @@ def array_access_parallelisable(self, loop_variable, var_info):
                               "does not depend on the loop "
```
Not from this PR but this error clearly means that the variable is being written to. However, I don't think there's any mention of this earlier in this routine. I think the routine name probably needs tweaking along with the docstring.

hiker replied:
Not sure - the test if it's written to is in the same function, so we can't just call it 'check_written_array'. I've added a list of warnings that might be raised in this function. Does this help?

arporter replied:
Ah, I must have missed that earlier check. Please just add a comment saying that we've already checked for this earlier.

hiker replied:
Done.

arporter replied:
Great, thanks.

arporter reviewed (changes requested):
Thanks for this improvement Joerg - I think it makes things a lot nicer.
That said, you'll see that I really struggled to understand the dependency code itself - I think some bits need renaming and/or docstrings extending.
I also see that you've added a 'tofortran' method and then use it to generate error messages.
I see two problems with this:
 1. We shouldn't be implementing code generation outside of the PSyIR and
 2. The PSyIR is language independent so we shouldn't be putting Fortran in the error message.
 
I admit that this might make wording the error message difficult but I think we'll have to stick to words like 'components' and 'structure'.

I've only looked at the documentation (which is very nice) and the code itself this time around. I'll look at tests etc. next time.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -73,7 +74,10 @@ def __init__(self, access_type, location, node, component_indices=None):
         self._access_type = access_type
         self._node = node
-        self.component_indices = component_indices
+        if not isinstance(component_indices, ComponentIndices):
```
Done.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -419,29 +420,54 @@ def next_location(self):
+            if component_indices is None:
+                component_indices = [[]] * len(signature)
+            elif isinstance(component_indices, list):
+                # 2. If the argument is a simple list (not a list of lists),
```
While I agree, we can't do this atm, there is still some code that provides strings. I've opened #1320 with a detailed list of where this happens, including the additional consistency checks that should be added. 
While it might be straightforward to wrap the strings in `Literal(...)` or `Reference(...)` calls, I would actually prefer to look at the way the PSyIR nodes are now created for the variables introduces by PSyclone, and use the instances that already exist (instead of recreating say a reference, that's never in the actual tree). It might lead to some code simplification.
Note I did not add #TODO 1320 in the sources, the list of files is in the ticket, and I don't want to touch all these other files for an issue that is not really related to this PR.

arporter replied:
That all sounds fine, thanks.

hiker replied:
Done.

arporter replied:
That's much clearer now, thanks.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+        :py:class:`psyclone.core.component_indices.ComponentIndices`, or \
+        None, [], a list or a list of lists of \
+        :py:class:`psyclone.psyir.nodes.Node`s
+
```
Oops - done.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+        elif isinstance(indices, list):
+            if all(isinstance(indx, list) for indx in indices):
+                # All elements of the indices are a list - so we
+                # got a list of lists:
```
#1320

arporter replied:
I bet you're glad you created that ticket now :-)

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+                # All elements of the indices are a list - so we
+                # got a list of lists:
+                self._component_indices = indices
+            elif all(not isinstance(indx, list) for indx in indices):
```
Yes, but #1320.

arporter replied:
Ah right.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+            elif all(not isinstance(indx, list) for indx in indices):
+                self._component_indices = [indices]
+            else:
+                raise InternalError("ComponentIndices: Invalid "
```
Done.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+            the variable is an array.
+        :rtype: bool
+        '''
+        return any(grp for grp in self._component_indices)
```
Done.

hiker commented on src/psyclone/core/signature.py:
```diff
@@ -99,6 +105,58 @@ def __getitem__(self, indx):
         return "%".join(self._signature)
 
+    # ------------------------------------------------------------------------
+    def to_fortran(self, component_indices):
```
OK, that might need some discussion. I would argue that you should get the terms that you provided in any error message. ATM we only allow Fortran as input language. And I don't see any way close to being as concise as what I have implemented.
My suggestion: we rename this function to (say) "to_input_language". We can then later, e.g. when we have other input language, query the language in the above function, and create output appropriate to the language currently used.

arporter replied:
In his work on adjointificaterisation, @rupertford has come up with a solution to this which at first I didn't like but I think does actually make sense: he passes an instance of a Visitor to the class and then, whenever he needs to generate a 'nice' error message, it's available to convert PSyIR into code. That way the class does not need to know what language is being used and we can still get nice error messages. If `Signature` kept a `Reference` (probably #1320 again?) then it would be really easy to do `my_visitor(signature.reference)` when required.

hiker replied:
OK, I will need to think about this, sounds like a good idea. But I should clarify one thing: I cannot store a Reference with the Signature (without significant refactoring). Simple example:
```
   a(i)%b(j) = 2
   a(j)%b(i) = 3
```
A VariablesAccessInfo instance will store both accesses for the same signature `(a,b)`, which is the key to get all accesses,. But I obviously cannot store a Reference with the signature here (which one should it be??) The Reference IS stored with each access. So e.g.:
```
all_access_info[Signature( (a,b) )] [0].node  --> a(i)%b(j)    # first access, assigning 2
all_access_info[Signature( (a,b) )] [1].node  --> a(j)%b(i)    # second access, assigning 3
```
So, once #1320 is done, I can switch to use the node information, but till that is done, I don't see any way to do that. But I can combine the signature with the ComponentIndices information to re-create the string that the psyir node will eventually produce.

hiker replied:
That indeed needs to be postponed till #1320, since at this stage we still might have strings, which a visitor cannot handle.  But I have added one TODO #1320 in this function.
*grin* I can't help myself from pointing out that psyGen.py also has a `bool_to_fortran` function, dynamo0p3 two `..._bound_fortran` functions as well ;)

arporter replied:
True, but they pre-date (I think/hope!) our move towards PSyIR and this is new code :-)

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -1878,7 +1878,7 @@ def kern_args(self, stub=False, var_accesses=None):
                     var_accesses.add_access(Signature(adj_face),
                                             AccessType.READ, self._kernel,
-                                            [[1]])
+                                            [1])
```
Yes, but #1320  (while this would be simple, I need to think about this a bit, it might be better to be ":" or so??). I don't think we have any kind of information about how this field is going to be used in the kernel in the first place, at least ":" would indicate that anything can be used.

arporter replied:
Oh, that's a point - I'd forgotten this was the LFRic API and therefore we won't (yet) necessarily have PSyIR to refer to. However, since it is the LFRic API then we may be able to deduce stuff from the rules.

hiker replied:
I assume you don't want me to do anything here? I do deduce stuff from the fact that it is LFRic: it is a read-only, 1-d array. I have to admit I am not sure how to get more, I can't know how this argument is used in the kernel, but the important point is that it is not dependent on the psy loop over all columns I think.

arporter replied:
Please add a comment to say that the `[1]` argument is to flag that the adjacent faces argument is a rank-1 array.

hiker replied:
Done, and added a todo #1320 as well (likely we might want to use ":" here ... though ... it's not actually clear to me ... do we have any information how the variables that PSyclone adds automatically are used???)

arporter replied:
Yes, I think we have quite a lot in the User Guide. If there's stuff missing then we should ask the scientists and flesh it out.

hiker commented on src/psyclone/gocean1p0.py:
```diff
@@ -1165,7 +1165,7 @@ def _record_stencil_accesses(self, signature, arg, var_accesses):
                     j_expr = GOKern._format_access("j", j, current_depth)
                     var_accesses.add_access(signature, arg.access,
-                                            self, [[i_expr, j_expr]])
+                                            self, [i_expr, j_expr])
```
#1320

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+        test for index values (e.g. `a(i,j)` and `a(i+3,j)`).
+        The caller needs to query the error messages managed by this object
+        to detect error(s), see `get_all_messages`.
+        This function returns the list of all accesses that use the loop
```
Done.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+        if not isinstance(var_infos, list):
+            all_accesses = var_infos.all_accesses
+        else:
+            all_accesses = []
```
Good idea, done.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+                    first_component_indices = component_indices
+                elif first_index != indx:
+                    # If a previously identified index location does not match
+                    # the current index location (e.g. a(i,j), and a(j,i) ),
```
We are, trust me, I am a professional  :) I've added a longish comment showing a loop, and which accesses we are looking at here. Hope that clarifies this a bit.

arporter replied:
Yes, I think I understand now. I think removing the `signature` argument was a good idea.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+                        "'{1}' is used differently: {2} and {3}."
+                        .format(signature.var_name,
+                                loop_variable.name,
+                                signature.to_fortran(first_component_indices),
```
Sea above.

hiker replied:
#1320, so won't fix this for now.

hiker reviewed (commented):
Hmm - I must have missed a few comments, I remember more. Will check, but that should be the majority of your comments.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -98,14 +102,14 @@ def change_read_to_write(self):
-        for the index expression used.
+        This function returns the list of accesses used for each component
+        as an instance of ComponentIndices. For example, `a(i)%b(j,k)%c`
+        will return `[ [i], [j, k], [] ]` as ComponentIndices. In the
```
Done.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -98,14 +102,14 @@ def change_read_to_write(self):
+        This function returns the list of accesses used for each component
+        as an instance of ComponentIndices. For example, `a(i)%b(j,k)%c`
+        will return `[ [i], [j, k], [] ]` as ComponentIndices. In the
+        case of a simple scalar variable `a` the component_indices will
```
Done.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -117,23 +121,20 @@ def component_indices(self, component_indices):
+            :py:class:`psyclone.core.component_indices.ComponentIndices`
 
-        :raises InternalError: if the component_indices is not a list of lists.
+        :raises InternalError: if the component_indices is not an instance \
```
Done.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -117,23 +121,20 @@ def component_indices(self, component_indices):
-            self._component_indices = [[]]
+
+        if not isinstance(component_indices, ComponentIndices):
+            raise InternalError("Index object in add_access must be "
```
Indeed :(

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -419,29 +420,54 @@ def next_location(self):
+            component_indices = ComponentIndices(component_indices)
+
+        if len(signature) != len(component_indices):
+            raise InternalError("Adding '{0}' as indices for '{1}', which "
```
Sounds good.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+class ComponentIndices(object):
+    '''This class stores index information for variable accesses. It stores
+    one index list for each component of a variable, e.g. for `a(i)%b(j)`
+    it would store `[ [i], [j] ]`. Even for scalar accesses an emtpy list
```
Done.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+    would have `[ [], [] ]`.
+
+    As a shortcut, the `indices` parameter can be None or an empty list
+    (which then creates the component indices as `[[]]`), a list `l` (which
```
I've reworded that - check if this is better. The issue is that the component indices does not know anything about the corresponding signature. In the access_info code that is handled (and indeed there it becomes the last component), here it is only a shortcut for non-structured variables, so I want to avoid 'last component' (while this is correct, it would be misleading).

arporter replied:
Yes, I think that's better - thanks.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+
+    :param indices: the indices from which to create this object.
+    :type indices: \
+        :py:class:`psyclone.core.component_indices.ComponentIndices`, or \
```
Indeed.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+                                    "list parameter '{0}'".format(indices))
+        else:
+            raise InternalError("Index object in ComponentIndices "
+                                "constructor must be a list or list of lists"
```
Done.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+    def __getitem__(self, indx):
+        '''Allows to use this class as a dictionary. If `indx` is an integer,
+        the list of indices for the specified component is returned. If `indx`
+        is a tuple (as returned from `iterate`), it will return the index
```
Done.

arporter replied:
I think we both got that wrong. It should probably say "...the PSyIR of the index for the specified component at the specified dimension."

hiker replied:
Indeed, I wrote nonsense.

hiker commented on src/psyclone/core/component_indices.py:
```diff
@@ -0,0 +1,142 @@
+        like `a(i)%b` with indices `[ [i], [] ]` would still be considered an
+        array.
+
+        :returns: if any of the variable components uses an index, i.e.\
```
Done (and I looked up the difference between if and whether ... maybe I learned something and can avoid this in the future ;) )

hiker commented on src/psyclone/domain/nemo/transformations/nemo_loop_fuse.py:
```diff
@@ -174,70 +177,16 @@ def validate_written_array(var_info1, var_info2, loop_variable):
-        # and 2nd dimension (j).
-        # TODO 1269: code duplicated with dependency_tools
-        found_dimension_index = None
+        from psyclone.psyir.tools import DependencyTools
```
Done.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
 
+    # -------------------------------------------------------------------------
+    def array_accesses_consistent(self, signature, loop_variable, var_infos):
+        '''Check if all accesses to the array have consistent usage of the
```
I had enough of this commenting nonsense ... I just remove signature from the argument list :)
(noticed when doing some work here that it is not actually required :) )

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+    # -------------------------------------------------------------------------
+    def array_accesses_consistent(self, signature, loop_variable, var_infos):
+        '''Check if all accesses to the array have consistent usage of the
+        loop variable in both VariableAccessInfo objects (used e.g. toverify
```
Sounds much better.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+        '''Check if all accesses to the array have consistent usage of the
+        loop variable in both VariableAccessInfo objects (used e.g. toverify
+        if loop can be fused by providing the access information of the each
+        loop in `var_info1` and `var_info2`).
```
I've reworded that - please check if this is better.

arporter replied:
Yes, I think it is better now, thanks. However, on looking at it again I think a method called `array_accesses_consistent` should return either `True` or `False`. As it stands, it returns a list of nodes and it's up to the user to check on the list of error messages to see whether the accesses are consistent or not. I suggest that the routine itself should do that.

hiker replied:
Done - I admit I don't like optional arguments to which information is added (for the `all_indices` information) ... no idea why tbh ...  but the code is indeed much nicer now, thanks!

arporter replied:
Thanks.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+                all_accesses = all_accesses + var_info.all_accesses
+
+        # The variable 'first_index' will store the index details of the
+        # first access to the variable ('index detail' being the information
```
I've reworded that all - hope it's better now.

arporter replied:
Ah yes, I get it now. Thanks.

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -108,6 +108,84 @@ def get_all_messages(self):
+        # a convenience value for the user
+        all_indices = []
+
+        # Loop over all the accesses of this variable
```
I've added more documentation, hope it is clearer now.

arporter replied:
Yes, much :-)

hiker commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -139,7 +217,7 @@ def _is_loop_suitable_for_parallel(self, loop, only_nested_loops=True):
-    def array_access_parallelisable(self, loop_variable, var_info):
+    def array_access_parallelisable(self, signature, loop_variable, var_info):
         '''Tries to determine if the access pattern for a variable
         given in var_info allows parallelisation along the variable
```
That argument is gone now.

hiker commented:
I've pushed quite a few things to #1320, and we might need a bit of a discussion about the `to_fortran` subroutine (unless you agree with just renaming it, and have a good idea for a name). My problem with the latter is that I can't think of a concise way to provide details to the user (but am happy for suggestions :) ).

arporter commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -133,28 +133,62 @@ def array_accesses_consistent(self, signature, loop_variable, var_infos):
+            consistency of accesses can be checked by querying the \
+            messages using `get_all_messages()`)
+        :rtype: list of :py:class:`psyclone.psyir.nodes.Node`
+
```
Please document the :raises:

hiker replied:
Done.

arporter reviewed (changes requested):
Again, I've only looked through the code changes this time. It's easier to understand now so thanks for updating comments and docstrings. I've suggested a solution to the `to_fortran` problem although it may be that it requires #1320 in order to properly address it?

hiker commented:
I believe I've addressed all issues - the one exception is the `to_fortran`, which indeed needs #1320 (since a visitor cannot handle a string argument ... I've tried that early in this PR ;) ). I hope we can discuss how we are going to handle multiple input languages in the telco tonight. Am happy to add another small PR before I can do #1320 to remove this function earlier.

hiker commented:
I've implemented the language support as discussed in the telco:
1. the `DependencyTools` take an optional `language_writer` argument (defaults to Fortran)
2. I've renamed `Signature.to_fortran` to `Signature.to_language`, which takes an optional language writer argument.
    At this stage, the `to_language` is still hard-coded to create Fortran-syntax (the index expression get converted with the language-dependent writer, but combining the indices into arrays is done in the subroutine, which still uses () and %. I've checked, and while it would only be limited amount of work to fully support C output (I had a hacked version running, but then noticed that it need some refactoring to avoid code duplication between C and Fortran), I've decided to open #1324 to track this ...  too much feature-creep otherwise. There are two long term solutions here: either we implement #1324 (which I might just do), and provide some useful functions in the writer that we could reuse (`FortranWriter._gen_dims()`, if we have the same function in C, we can be much more language-independent, then we only need to make % vs . configurable)). Or we implement #1320 first, in which case we don't need that function at all (as far as I can see atm, since then we will always have the right `Reference` node as an index, and not a string anymore), and we can use the language writer directly to convert the `Reference` used in an access. Both are required, #1320 might be more work (due to at very least LFRic relying on strings at this stage).
3. I've documented this behaviour in the docs (and referenced both tickets ... that's a bit odd admittedly, but as stated above, it's hard to decide what will come first).

My personal expectation is to do #1324 first, which will allow us to create the output string language-independent, and then when #1320 comes, the whole `Signature.to_language` function will be removed and replaced with `language_writer(this_access.node)`. Still, #1324 will add support for structures to the C writer and is obviously useful to implement.

hiker commented:
Ready to go for next review (there is one comment I cannot answer ... about array_access_consistent, and that's done). Note my previous comment about what I have pushed to other tickets to get this one over the line.

arporter commented on doc/developer_guide/psyir_symbols.rst:
```diff
@@ -669,3 +669,9 @@ can be parallelised::
               for message in dt.get_all_messages():
                   print(message)
+
+.. note:: While the dependency tools constructor take an argument
```
s/take/takes/

hiker replied:
Done.

arporter commented on src/psyclone/core/access_info.py:
```diff
@@ -430,8 +430,17 @@ def next_location(self):
+        `component_indices=[i,j]`, it will create `[[], [i,j]` as component
+        indices, indicating that no index is used in the first component `a`.
+        If the access is supposed to be for `a(i)%b(j)`, then the
+        `component_indices` argument must be specified as list of lists,
```
s/as list/as a list/

hiker replied:
Done.

arporter commented on src/psyclone/core/access_info.py:
```diff
@@ -430,8 +430,17 @@ def next_location(self):
-        ComponentIndices, it is used to construct an instance. Therefore it
-        can be None, a list or a list of lists of PSyIR nodes.
+        `ComponentIndices`, it is used to construct an instance. Therefore it
+        can be None, a list or a list of lists of PSyIR nodes. In case of
```
s/In case/In the case/

hiker replied:
Done.

arporter commented on src/psyclone/core/access_info.py:
```diff
@@ -430,8 +430,17 @@ def next_location(self):
-        can be None, a list or a list of lists of PSyIR nodes.
+        `ComponentIndices`, it is used to construct an instance. Therefore it
+        can be None, a list or a list of lists of PSyIR nodes. In case of
+        a list of lists, this will be unmodified used to construct the
```
s/unmodified used/used unmodified/

hiker replied:
Done.

arporter commented on src/psyclone/core/access_info.py:
```diff
@@ -430,8 +430,17 @@ def next_location(self):
+        a list of lists, this will be unmodified used to construct the
+        ComponentIndices structures. If it is a simple list, it is assumed
+        that it contains the indices used in accessing the last component
+        of the signature. For example, for `a%b` with
```
Good description - thanks :-)

arporter commented on src/psyclone/core/signature.py:
```diff
@@ -106,14 +106,19 @@ def __str__(self):
-    def to_fortran(self, component_indices):
-        '''Converts this signature back to a Fortran expression using the
-        provided component indices.
+    def to_language(self, component_indices, language_writer=None):
```
Since `FortranWriter` is stateless, I think you can just have `language_writer=FortranWriter()` here and drop the `if xxx is None` below.

hiker replied:
I can't here, I am suffering from a circular import if I import the FortranWriter outside of the functions.

arporter replied:
That's a shame.

arporter commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -54,19 +54,28 @@ class DependencyTools(object):
-
-    def __init__(self, loop_types_to_parallelise=None):
+    def __init__(self, loop_types_to_parallelise=None,
+                 language_writer=None):
```
Again, I think `language_writer=FortranWriter()` here is fine.

hiker replied:
It worked here  (or the other way round) - it worked in one place and I've replaced it.

arporter commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -110,19 +119,19 @@ def get_all_messages(self):
-        variable. This is an additional convenient result that can simplify
-        further tests (but can also be ignored).
+        the loop variable. This can be used e.g. to verify whether two loops
+        may be fused by providing the access information of the each loop in
```
s/of the each/of each/

hiker replied:
Indeed.

arporter commented on src/psyclone/tests/core/signature_test.py:
```diff
@@ -173,3 +178,64 @@ def test_signature_comparison():
+    assert sig.to_language(comp) == "a(j)"
+
+    # Test error condition if number of components in signature does not
+    # math the number of indices in ComponentIndices
```
"match"

hiker replied:
Done.

arporter commented on src/psyclone/tests/core/signature_test.py:
```diff
@@ -173,3 +178,64 @@ def test_signature_comparison():
+    comp = ComponentIndices([[1], [2]])
+    with pytest.raises(InternalError) as err:
+        sig.to_language(comp)
+    assert "Signature a has 1 components, but component_indices [[1], [2]] " \
```
Quotes around the signature, i.e. "Signature 'a' has ...."

hiker replied:
Done, and I noticed one or two more messages without '' - not saying I fixed them all, but it's better now :)

arporter replied:
Thanks Joerg.

arporter commented on src/psyclone/tests/core/signature_test.py:
```diff
@@ -173,3 +178,64 @@ def test_signature_comparison():
         in str(err.value)
+
+
+def test_to_fortran():
```
Maybe rename to "test_to_language_fortran"?

hiker replied:
Done.

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -172,25 +175,115 @@ def test_arrays_parallelise(parser):
+                                     end do
+                                 end do
+                                 end program test''')
+    prog = parser(reader)
```
I think you can use the `fortran_reader` fixture here. Just add it as an argument to the test and then you can do:

    psyir = fortran_reader('''program ... ...end program''')

hiker replied:
I can't: if I do this, I don't get loops created by Nemo API, and as a result, these loops do not have a type (latitude, ... etc) - and will therefore be rejected by the dependency tools. I could of course disable the loop type test, but I think it's useful to have that.

arporter replied:
Oh yes, I forgot about that, sorry. (You could raise the PSyIR provided by the fortran reader but I'm happy for you to leave it as it is.)

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -172,25 +175,115 @@ def test_arrays_parallelise(parser):
+                                                 var_info1[sig_b],
+                                                 var_info1[sig_c]])
+    assert "Inconsistent signature provided in 'array_accesses_consistent'. " \
+           "Expected all access to be for 'a', but also got 'b,c'."\
```
s/access/accesses/

hiker replied:
Done.

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -172,25 +175,115 @@ def test_arrays_parallelise(parser):
+                                                     [a_access_1st_loop,
+                                                      a_access_2nd_loop],
+                                                     all_ind)
+    assert consistent
```
Please add `is True` just to make sure that the return value is what we expect.

hiker replied:
OK, that's new to me. I've updated a lot of tests to use `is` now.

arporter replied:
Thanks Joerg - it's almost paranoia but it doesn't hurt to be careful :-)

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -172,25 +175,115 @@ def test_arrays_parallelise(parser):
+                                                     [a_access_1st_loop,
+                                                      a_access_3rd_loop],
+                                                     all_ind)
+    assert not consistent
```
Please change to `is False`.

hiker replied:
Done.

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -172,25 +175,115 @@ def test_arrays_parallelise(parser):
+    assert not consistent
+    assert len(dep_tools.get_all_messages()) == 1
+    assert "Variable 'a' is written to and the loop variable 'jj' is used " \
+           "differently: a(ji,jj) and a(jj,ji)." \
```
s/is used differently/is used in different index locations/ perhaps?

arporter replied:
(I can see the value of having actual Fortran in the error message here though.)

hiker replied:
I'm glad that you agree and this wasn't all wasted :)

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -172,25 +175,115 @@ def test_arrays_parallelise(parser):
+                                                     all_ind)
+    assert not consistent
+    assert len(dep_tools.get_all_messages()) == 1
+    assert "Variable 'a' is written to and the loop variable 'jj' is used " \
```
For consistency with other tests, please could you enclose the condition for the assert within parentheses rather than using `\`.

hiker replied:
That is new to me, too. Didn't we a while ago say to get rid of the `()` for `asserts`?? I admitted I did a quick `grep`, and was amazed how many `assert ()` we have, and only a few `\`. So I've fixed this, and a lot of others as well that I noticed.

arporter replied:
Yes, I think it's the PEP recommendation to use `()` rather than `\`. The only thing to be careful of with asserts is that you put the parenthesis around the whole expression otherwise you may just be getting it to tell you that your comparison string is not empty ;-)

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -246,24 +339,28 @@ def test_scalar_parallelise(declaration, variable, parser):
+    dep_tools = DependencyTools(["levels", "lat"],
+                                language_writer=FortranWriter())
+
+    # Test if supplying the loop name as symbol also works:
```
What does "also works" mean here? I thought we now always have to supply a symbol?

hiker replied:
Good point, I forgot that I had changed this consistently everywhere. I then decided to also update the information that declared these arguments in several places to be a `string` still 😉

arporter commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -293,33 +390,33 @@ def test_derived_type(parser):
+        can_loop_be_parallelised(loops[1],
                                  signatures_to_ignore=[Signature(("a", "b"))])
     assert not parallel
     assert len(dep_tools.get_all_messages()) == 1
```
Not this PR but probably worth checking that the one message is the right one!

hiker replied:
Indeed - I believe I've fixed all tests that did not test for the details of the message.

arporter reviewed (changes requested):
Looking good now Joerg.
I've now looked at all the code and all the tests.
Looking at the tests made me realise that I don't like the `ComponentIndices.get()` method name so I have proposed changing that. There are a few other minor things that the tests made me notice too.
I'll check for pylint and pycodestyle next time.
All tests and examples are OK with compilation.
Updated docs build fine.

hiker commented on src/psyclone/core/access_info.py:
```diff
@@ -250,20 +260,20 @@ def all_accesses(self):
 
     def add_access_with_location(self, access_type, location, node,
-                                 component_indices=None):
+                                 component_indices):
```
Yes ;) I thought about that. I decided to remove this, because this simple class is only used internally (by `VariablesAccessesInfo`). When we automatically create component_indices in `VariablesAccessInfos`, we have access to the signature, i.e. we can make sure to create a consistent `ComponentIndices` object. In this class however, we don't have access to the signature, and therefore we can't guarantee that we will create a consistent set (e.g.: 'field%global%startx' would create `[[]]``, not `[[][][]]`.
Since the class will not be created by an application (I can't see a good reason, an application will always use `VariablesAccessInfos` as far as I can see), I disabled the default to guarantee we always have consistent info - and accepted the fact that I needed to be more verbose in the tests.

arporter replied:
OK, that's good enough for me :-)

hiker commented on src/psyclone/tests/core/access_info_test.py:
```diff
@@ -167,24 +181,28 @@ def test_variable_access_info_read_write():
     node = Node()
-    vai.add_access_with_location(AccessType.READ, 2, node)
+    vai.add_access_with_location(AccessType.READ, 2, node,
+                                 component_indices=None)
```
As written above, I prefer not to (but I should have mentioned before: if you insist, I will).

hiker commented on src/psyclone/tests/core/access_type_test.py:
```diff
@@ -62,6 +62,16 @@ def test_api_specific_name():
     assert AccessType.READ.api_specific_name() == "gh_read"
     assert AccessType.READWRITE.api_specific_name() == "gh_readwrite"
     assert AccessType.SUM.api_specific_name() == "gh_sum"
+    assert AccessType.get_valid_reduction_modes() == [AccessType.SUM]
```
Done.

hiker commented on src/psyclone/tests/core/component_indices_test.py:
```diff
@@ -0,0 +1,104 @@
+    '''
+
+    component_indices = ComponentIndices([["a", "b"], ["c"]])
+    correct_index_pairs = [(0, 0), (0, 1), (1, 0)]
```
I've added tests that raise appropriate IndexErrors with details about which index is wrong.

hiker commented on src/psyclone/tests/core/access_info_test.py:
```diff
@@ -248,6 +273,43 @@ def test_variables_access_info():
+# -----------------------------------------------------------------------------
+def test_component_indices_auto_extension():
+    '''To make it more convenient for the user certain combinations of
+    signature and component_indices in the add_location vall will
```
Indeed.

hiker commented on src/psyclone/tests/core/access_info_test.py:
```diff
@@ -248,6 +273,43 @@ def test_variables_access_info():
+    # This should auto-extent the component indices,
+    # since they are specified as a simple list:
+    var_accesses.add_access(sig, AccessType.READ, node, ["i", "j"])
+    assert var_accesses[sig][0].component_indices.get() == [[], [], ["i", "j"]]
```
Properties ... I don't like ... well, done :P

arporter replied:
Thanks :)

hiker commented on src/psyclone/tests/core/access_info_test.py:
```diff
@@ -248,6 +273,43 @@ def test_variables_access_info():
+    with pytest.raises(InternalError) as err:
+        var_accesses.add_access(sig, AccessType.READ, node, [["i", "j"]])
+    assert "Cannot add '[['i', 'j']]' as indices for 'a%b%c' as the number "\
+           "of components do not match." in str(err.value)
```
Done.

arporter replied:
Thanks, I think that's helpful.

hiker commented on src/psyclone/tests/core/component_indices_test.py:
```diff
@@ -0,0 +1,104 @@
+    '''
+
+    component_indices = ComponentIndices()
+    assert component_indices.get() == [[]]
```
Done.

hiker reviewed (commented):
Github asks me to submit my started review ... again, not sure why, but I better type something here before github loses it all :)

hiker commented:
OK, should be ready for the next review.

arporter commented on src/psyclone/core/component_indices.py:
```diff
@@ -116,8 +116,16 @@ def __getitem__(self, indx):
         :rtype: list of :py:class:`psyclone.psyir.nodes.Node`, or \
             :py:class:`psyclone.psyir.nodes.Node`
 
+        :raises IndexError:
```
Please complete :-)

hiker replied:
Now that is really embarrassing (well, I originally didn't add detailed error messages, so I added this, then tests, then pylint, then ... forgot about finishing it :( )

arporter commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -406,9 +407,11 @@ def can_loop_be_parallelised(self, loop, loop_variable=None,
+            parallelised. If not specified, the loop variable of the loop \
+            is used.
+        :type loop_variable: \
+            :py:class:`psyclone.psyir.symbol.datasymbol.DataSymbol`
```
Just `psyclone.psyir.symbols.DataSymbol` I think?

hiker replied:
Ah yes. I was actually wondering about this in some other cases, but can't remember where now ... might have been different PR.

arporter commented on src/psyclone/tests/core/component_indices_test.py:
```diff
@@ -102,3 +102,18 @@ def test_iterating():
+
+# -----------------------------------------------------------------------------
+def test_component_indices_getitem_exceptions():
+    '''Tests useful error messages are provides if a tuple is provided that
```
s/provides/provided/

hiker replied:
Done.

arporter commented on src/psyclone/tests/core/component_indices_test.py:
```diff
@@ -102,3 +102,18 @@ def test_iterating():
+    is out of bounds.
+    '''
+    component_indices = ComponentIndices([["a", "b"], ["c"]])
+    with pytest.raises(IndexError) as err:
```
Please could you extend this to cover the lower-bound cases too (e.g. attempting to access `-1`).

hiker replied:
Done.

arporter reviewed (changes requested):
We're down in the nitty-gritty now.
All major things have been addressed.
Tests and examples still all OK with compilation.
Updated documentation builds fine.

hiker commented:
Ready to go.

arporter reviewed (approved):
All requested changes have been made.
All tests and examples still OK with compilation.
All docs build fine.
This is ready to be merged.

arporter commented:
#1323 is ahead of this one. Once that's on, I shall merge this one.
